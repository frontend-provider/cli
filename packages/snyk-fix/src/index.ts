import * as debugLib from 'debug';
import { loadPlugin } from './plugins/load-plugin';
import { FixHandlerResult } from './plugins/types';

import { EntityToFix } from './types';

const debug = debugLib('snyk-fix:main');

export async function fix(
  entities: EntityToFix[],
): Promise<FixHandlerResult[]> {
  debug(`Requested to fix ${entities.length} projects.`);
  const entitiesPerType = groupEntitiesPerType(entities);
  const allResults: FixHandlerResult[] = [];
  // TODO: pMap this?
  for (const type of Object.keys(entitiesPerType)) {
    const handler = loadPlugin(type);
    const results = await handler(entitiesPerType[type]);
    allResults.push(results);
  }

  return allResults;
}

function groupEntitiesPerType(
  entities: EntityToFix[],
): {
  [type: string]: EntityToFix[];
} {
  const entitiesPerType: {
    [type: string]: EntityToFix[];
  } = {};
  for (const entity of entities) {
    const type = entity.scanResult.identity.type;
    if (entitiesPerType[type]) {
      entitiesPerType[type].push(entity);
      continue;
    }
    entitiesPerType[type] = [entity];
  }
  return entitiesPerType;
}
