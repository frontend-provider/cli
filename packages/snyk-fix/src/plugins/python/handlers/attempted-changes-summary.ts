import {
  DependencyPins,
  FixChangesError,
  FixChangesSuccess,
  FixChangesSummary,
} from '../../../types';

export function generateFailedChanges(
  attemptedUpdates: string[],
  pins: DependencyPins,
  error: Error,
  command?: string,
): FixChangesError[] {
  const changes: FixChangesError[] = [];
  for (const pkgAtVersion of Object.keys(pins)) {
    const pin = pins[pkgAtVersion];
    if (
      !attemptedUpdates
        .map((update) => update.replace('==', '@'))
        .includes(pin.upgradeTo)
    ) {
      continue;
    }
    const updatedMessage = pin.isTransitive ? 'pin' : 'upgrade';
    const newVersion = pin.upgradeTo.split('@')[1];
    const [pkgName, version] = pkgAtVersion.split('@');

    changes.push({
      success: false,
      reason: error.message,
      userMessage: `Failed to ${updatedMessage} ${pkgName} from ${version} to ${newVersion}`,
      tip: command ? `Try running \`${command}\`` : undefined,
      issueIds: pin.vulns,
      from: pkgAtVersion,
      to: `${pkgName}@${newVersion}`,
    });
  }
  return changes;
}

export function generateSuccessfulChanges(
  appliedUpgrades: string[],
  pins: DependencyPins,
): FixChangesSuccess[] {
  const changes: FixChangesSuccess[] = [];
  for (const pkgAtVersion of Object.keys(pins)) {
    const pin = pins[pkgAtVersion];
    if (
      !appliedUpgrades
        .map((upgrade) => upgrade.replace('==', '@'))
        .includes(pin.upgradeTo)
    ) {
      continue;
    }
    const updatedMessage = pin.isTransitive ? 'Pinned' : 'Upgraded';
    const newVersion = pin.upgradeTo.split('@')[1];
    const [pkgName, version] = pkgAtVersion.split('@');

    changes.push({
      success: true,
      userMessage: `${updatedMessage} ${pkgName} from ${version} to ${newVersion}`,
      issueIds: pin.vulns,
      from: pkgAtVersion,
      to: `${pkgName}@${newVersion}`,
    });
  }
  return changes;
}

export function isSuccessfulChange(
  change: FixChangesSummary,
): change is FixChangesError {
  return change.success === true;
}
