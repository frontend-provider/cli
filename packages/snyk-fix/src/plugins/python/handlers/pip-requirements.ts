import * as debugLib from 'debug';
import { EntityToFix } from '../../../types';
import { updateDependencies } from './update-dependencies';

const debug = debugLib('snyk-fix:python:requirements.txt');

export async function pipRequirementsTxt(
  entities: EntityToFix[],
): Promise<{
  succeeded: EntityToFix[];
  failed: EntityToFix[];
  skipped: EntityToFix[];
}> {
  debug(`Preparing to fix ${entities.length} Python requirements.txt projects`);
  const handlerResult: {
    succeeded: EntityToFix[];
    failed: EntityToFix[];
    skipped: EntityToFix[];
  } = {
    succeeded: [],
    failed: [],
    skipped: [],
  };
  for (const entity of entities) {
    try {
      if (await notSupported(entity)) {
        handlerResult.skipped.push(entity);
        continue;
      }
      const fixedEntity = await fixIndividualRequirementsTxt(entity);
      handlerResult.succeeded.push(fixedEntity);
    } catch (e) {
      console.error(e); // TODO: use spinner & propagate error back
      handlerResult.failed.push(entity);
    }
  }
  return handlerResult;
}

export async function notSupported(entity: EntityToFix): Promise<boolean> {
  const remediationData = entity.testResult.remediation;
  if (!remediationData) {
    return true;
  }

  if (await containsRequireDirective(entity)) {
    return true;
  }
  return false;
}

/* Requires like -r, -c are not supported at the moment, as multiple files
 * would have to be identified and fixed together
 * https://pip.pypa.io/en/stable/reference/pip_install/#options
 */
async function containsRequireDirective(entity: EntityToFix): Promise<boolean> {
  const REQUIRE_PATTERN = /^[^\S\n]*-(r|c)\s+.+/;
  // TODO: fix the non null assertion here
  const fileName = entity.scanResult.identity.targetFile!;
  const requirementsTxt = await entity.workspace.readFile(fileName);
  const match = REQUIRE_PATTERN.exec(requirementsTxt);
  if (match && match.length > 1) {
    return true;
  }
  return false;
}

// TODO: optionally verify the deps install
export async function fixIndividualRequirementsTxt(
  entity: EntityToFix,
): Promise<EntityToFix> {
  const fileName = entity.scanResult.identity.targetFile;
  const remediationData = entity.testResult.remediation;
  if (!remediationData) {
    throw new Error('Fixing is not available without remediation data');
  }
  if (!fileName) {
    // TODO: is this possible?
    throw new Error('Requirements file name required');
  }
  const requirementsTxt = await entity.workspace.readFile(fileName);
  // TODO: allow handlers per fix type (later also strategies or combine with strategies)
  const fixedRequirementsTxt = updateDependencies(
    requirementsTxt,
    remediationData.pin,
  );
  await entity.workspace.writeFile(fileName, fixedRequirementsTxt);
  // TODO: generate fixes per file + failed per file to generate clear output later
  return entity;
}
