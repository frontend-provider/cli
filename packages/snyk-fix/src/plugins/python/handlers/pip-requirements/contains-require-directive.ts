/* Requires like -r, -c are not supported at the moment, as multiple files
 * would have to be identified and fixed together
 * https://pip.pypa.io/en/stable/reference/pip_install/#options
 */
export async function containsRequireDirective(
  requirementsTxt: string,
): Promise<{ containsRequire: boolean; matches: RegExpMatchArray[] }> {
  const allMatches: RegExpMatchArray[] = [];
  const REQUIRE_PATTERN = new RegExp(/^[^\S\n]*-(r|c)\s+(.+)/, 'gm');
  const matches = getAllMatchedGroups(REQUIRE_PATTERN, requirementsTxt);
  for (const match of matches) {
    if (match && match.length > 1) {
      allMatches.push(match);
    }
  }
  return { containsRequire: allMatches.length > 0, matches: allMatches };
}

function getAllMatchedGroups(re: RegExp, str: string) {
  const groups: RegExpExecArray[] = [];
  let match: RegExpExecArray | null;
  while ((match = re.exec(str))) {
    groups.push(match);
  }

  return groups;
}
