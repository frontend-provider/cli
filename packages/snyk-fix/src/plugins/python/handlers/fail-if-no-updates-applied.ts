import * as debugLib from 'debug';

import { NoFixesCouldBeAppliedError } from '../../../lib/errors/no-fixes-applied';
import { FixChangesError, FixChangesSummary } from '../../../types';
import { isSuccessfulChange } from './attempted-changes-summary';

const debug = debugLib('snyk-fix:python:ensure-changes-applied');

export function failIfNoUpdatesApplied(changes: FixChangesSummary[]) {
  if (!changes.length) {
    throw new NoFixesCouldBeAppliedError();
  }
  if (!changes.some((c) => isSuccessfulChange(c))) {
    debug('Manifest has not changed as no changes got applied!');
    // throw the first error tip since 100% failed, they all failed with the same
    // error
    const { reason, tip } = changes[0] as FixChangesError;
    throw new NoFixesCouldBeAppliedError(reason, tip);
  }
}
