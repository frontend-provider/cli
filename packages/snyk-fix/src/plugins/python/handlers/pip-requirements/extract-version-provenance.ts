import * as path from 'path';
import * as debugLib from 'debug';

import {
  ParsedRequirements,
  parseRequirementsFile,
} from './update-dependencies/requirements-file-parser';
import { Workspace } from '../../../../types';
import { containsRequireDirective } from './contains-require-directive';

export interface PythonProvenance {
  [fileName: string]: ParsedRequirements;
}

const debug = debugLib('snyk-fix:python:extract-version-provenance');

export async function extractProvenance(
  workspace: Workspace,
  dir: string,
  fileName: string,
  provenance: PythonProvenance = {},
): Promise<PythonProvenance> {
  const requirementsFileName = path.join(dir, fileName);
  const requirementsTxt = await workspace.readFile(requirementsFileName);
  provenance = {
    ...provenance,
    [fileName]: parseRequirementsFile(requirementsTxt),
  };
  const { containsRequire, matches } = await containsRequireDirective(
    requirementsTxt,
  );
  if (containsRequire) {
    for (const match of matches) {
      const requiredFilePath = match[2];
      if (provenance[requiredFilePath]) {
        debug('Detected recursive require directive, skipping');
        continue;
      }

      provenance = {
        ...provenance,
        ...(await extractProvenance(
          workspace,
          dir,
          requiredFilePath,
          provenance,
        )),
      };
    }
  }
  return provenance;
}
