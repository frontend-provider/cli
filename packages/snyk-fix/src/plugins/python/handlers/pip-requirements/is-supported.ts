import * as chalk from 'chalk';

import { EntityToFix, WithUserMessage } from '../../../../types';
import { containsRequireDirective } from './contains-require-directive';

interface Supported {
  supported: true;
}

interface NotSupported {
  supported: false;
  reason: string;
}

export function projectTypeSupported(
  res: Supported | NotSupported,
): res is Supported {
  return !('reason' in res);
}

export async function isSupported(
  entity: EntityToFix,
): Promise<Supported | NotSupported> {
  const remediationData = entity.testResult.remediation;

  if (!remediationData) {
    return { supported: false, reason: 'No remediation data available' };
  }
  if (!remediationData.pin || Object.keys(remediationData.pin).length === 0) {
    return {
      supported: false,
      reason: 'There is no actionable remediation to apply',
    };
  }

  // TODO: fix the non null assertion here
  let requirementsTxt;
  try {
    const fileName = entity.scanResult.identity.targetFile!;
    requirementsTxt = await entity.workspace.readFile(fileName);
  } catch (e) {
    return {
      supported: false,
      reason: e.message,
    };
  }

  const { containsRequire } = await containsRequireDirective(requirementsTxt);
  if (containsRequire) {
    return {
      supported: false,
      reason: `Requirements with ${chalk.bold('-r')} or ${chalk.bold(
        '-c',
      )} directive are not yet supported`,
    };
  }

  return { supported: true };
}

export async function partitionByFixable(
  entities: EntityToFix[],
): Promise<{
  skipped: Array<WithUserMessage<EntityToFix>>;
  fixable: EntityToFix[];
}> {
  const fixable: EntityToFix[] = [];
  const skipped: Array<WithUserMessage<EntityToFix>> = [];
  for (const entity of entities) {
    const isSupportedResponse = await isSupported(entity);
    if (projectTypeSupported(isSupportedResponse)) {
      fixable.push(entity);
      continue;
    }
    skipped.push({
      original: entity,
      userMessage: isSupportedResponse.reason,
    });
  }
  return { fixable, skipped };
}
