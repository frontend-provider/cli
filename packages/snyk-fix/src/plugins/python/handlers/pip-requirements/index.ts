import * as debugLib from 'debug';
import * as pathLib from 'path';

import {
  DependencyPins,
  EntityToFix,
  FixChangesSummary,
  FixOptions,
  RemediationChanges,
  Workspace,
} from '../../../../types';
import { PluginFixResponse } from '../../../types';
import { updateDependencies } from './update-dependencies';
import { MissingRemediationDataError } from '../../../../lib/errors/missing-remediation-data';
import { MissingFileNameError } from '../../../../lib/errors/missing-file-name';
import { partitionByFixable } from './is-supported';
import { NoFixesCouldBeAppliedError } from '../../../../lib/errors/no-fixes-applied';
import {
  extractProvenance,
  PythonProvenance,
} from './extract-version-provenance';
import {
  ParsedRequirements,
  parseRequirementsFile,
  Requirement,
} from './update-dependencies/requirements-file-parser';

const debug = debugLib('snyk-fix:python:requirements.txt');

export async function pipRequirementsTxt(
  entities: EntityToFix[],
  options: FixOptions,
): Promise<PluginFixResponse> {
  debug(`Preparing to fix ${entities.length} Python requirements.txt projects`);
  const handlerResult: PluginFixResponse = {
    succeeded: [],
    failed: [],
    skipped: [],
  };

  const { fixable, skipped } = await partitionByFixable(entities);
  handlerResult.skipped.push(...skipped);

  for (const entity of fixable) {
    try {
      const { changes } = await applyAllFixes(
        entity,
        // dir,
        // base,
        // remediation,
        // provenance,
        options,
      );
      if (!changes.length) {
        debug('Manifest has not changed!');
        throw new NoFixesCouldBeAppliedError();
      }
      handlerResult.succeeded.push({ original: entity, changes });
    } catch (e) {
      handlerResult.failed.push({ original: entity, error: e });
    }
  }
  return handlerResult;
}

export function getRequiredData(
  entity: EntityToFix,
): {
  remediation: RemediationChanges;
  targetFile: string;
  workspace: Workspace;
} {
  const { remediation } = entity.testResult;
  if (!remediation) {
    throw new MissingRemediationDataError();
  }
  const { targetFile } = entity.scanResult.identity;
  if (!targetFile) {
    throw new MissingFileNameError();
  }
  const { workspace } = entity;
  if (!workspace) {
    throw new NoFixesCouldBeAppliedError();
  }
  return { targetFile, remediation, workspace };
}

// TODO: optionally verify the deps install
export async function fixIndividualRequirementsTxt(
  workspace: Workspace,
  dir: string,
  entryFileName: string,
  fileName: string,
  remediation: RemediationChanges,
  parsedRequirements: ParsedRequirements,
  options: FixOptions,
  directUpgradesOnly: boolean,
): Promise<{ changes: FixChangesSummary[]; appliedRemediation: string[] }> {
  const fullFilePath = pathLib.join(dir, fileName);
  const { updatedManifest, changes, appliedRemediation } = updateDependencies(
    parsedRequirements,
    remediation.pin,
    directUpgradesOnly,
    pathLib.join(dir, entryFileName) !== fullFilePath ? fileName : undefined,
  );
  if (!options.dryRun && changes.length > 0) {
    debug('Writing changes to file');
    await workspace.writeFile(pathLib.join(dir, fileName), updatedManifest);
  } else {
    debug('Skipping writing changes to file in --dry-run mode');
  }

  return { changes, appliedRemediation };
}

export async function applyAllFixes(
  entity: EntityToFix,
  options: FixOptions,
): Promise<{ changes: FixChangesSummary[] }> {
  const { remediation, targetFile: entryFileName, workspace } = getRequiredData(
    entity,
  );
  const { dir, base } = pathLib.parse(entryFileName);
  const provenance = await extractProvenance(workspace, dir, base);
  const upgradeChanges: FixChangesSummary[] = [];
  const appliedUpgradeRemediation: string[] = [];
  for (const fileName of Object.keys(provenance)) {
    const skipApplyingPins = true;
    const { changes, appliedRemediation } = await fixIndividualRequirementsTxt(
      workspace,
      dir,
      base,
      fileName,
      remediation,
      provenance[fileName],
      options,
      skipApplyingPins,
    );
    appliedUpgradeRemediation.push(...appliedRemediation);
    // what if we saw the file before and already fixed it?
    upgradeChanges.push(...changes);
  }
  // now do left overs as pins + add tests
  const requirementsTxt = await workspace.readFile(entryFileName);

  const toPin: RemediationChanges = filterOutAppliedUpgrades(
    remediation,
    appliedUpgradeRemediation,
  );
  const directUpgradesOnly = false;
  const { changes: pinnedChanges } = await fixIndividualRequirementsTxt(
    workspace,
    dir,
    base,
    base,
    toPin,
    parseRequirementsFile(requirementsTxt),
    options,
    directUpgradesOnly,
  );

  return { changes: [...upgradeChanges, ...pinnedChanges] };
}

function filterOutAppliedUpgrades(
  remediation: RemediationChanges,
  appliedRemediation: string[],
): RemediationChanges {
  const pinRemediation: RemediationChanges = {
    ...remediation,
    pin: {}, // delete the pin remediation so we can add only not applied
  };
  const pins = remediation.pin;
  const lowerCasedAppliedRemediation = appliedRemediation.map((i) =>
    i.toLowerCase(),
  );
  for (const pkgAtVersion of Object.keys(pins)) {
    if (!lowerCasedAppliedRemediation.includes(pkgAtVersion.toLowerCase())) {
      pinRemediation.pin[pkgAtVersion] = pins[pkgAtVersion];
    }
  }
  return pinRemediation;
}
