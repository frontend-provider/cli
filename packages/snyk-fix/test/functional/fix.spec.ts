import * as fs from 'fs';

import * as snykFix from '../../src';
import { generateEntityToFix } from '../helpers/generate-entity-to-fix';
import * as pipRequirementsFix from '../../src/plugins/python/handlers/pip-requirements';
describe('Snyk fix', () => {
  it('Snyk fix throws error when called with unsupported type', () => {
    // read data from console.error
    let stdoutMessages = '';
    const stubConsoleError = (msg: string) => (stdoutMessages += msg);
    const origConsoleLog = console.error;
    console.error = stubConsoleError;

    const projectTestResult = generateEntityToFix('npm', 'package.json');
    expect(
      snykFix.fix([projectTestResult]),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      '"Provided scan type is not supported"',
    );
    expect(stdoutMessages).toMatchSnapshot();
    // restore original console.error
    console.log = origConsoleLog;
  });
  it('Snyk fix returns results for supported type', async () => {
    const projectTestResult = generateEntityToFix('pip', 'requirements.txt');
    const res = await snykFix.fix([projectTestResult]);
    expect(res).toMatchSnapshot();
  });

  it('Snyk fix returns results for supported & unsupported type', async () => {
    const projectTestResult = generateEntityToFix('pip', 'requirements.txt');
    const pipfileProjectTestResult = generateEntityToFix('pip', 'Pipfile');
    const res = await snykFix.fix([
      projectTestResult,
      pipfileProjectTestResult,
    ]);
    expect(res).toMatchSnapshot();
  });

  it('Snyk fix returns results as expected', async () => {
    const txtProdProjectTestResult = generateEntityToFix('pip', 'prod.txt');
    const txtDevProjectTestResult = generateEntityToFix('pip', 'dev.txt');
    const pipfileProjectTestResult = generateEntityToFix('pip', 'Pipfile');
    const res = await snykFix.fix([
      txtDevProjectTestResult,
      txtProdProjectTestResult,
      pipfileProjectTestResult,
    ]);
    expect(res.exceptionsByScanType).toEqual({});
    expect(Object.keys(res.resultsByPlugin)).toHaveLength(1);
    expect(Object.keys(res.resultsByPlugin)[0]).toEqual('python');
    // skipped unsupported
    expect(res.resultsByPlugin.python.skipped).toHaveLength(1);
    expect(res.resultsByPlugin.python.skipped[0]).toEqual(
      pipfileProjectTestResult,
    );

    // first *.txt throws because of the mock above
    expect(res.resultsByPlugin.python.failed).toHaveLength(0);
    // TODO: only 1 should succeed once implemented the fix
    expect(res.resultsByPlugin.python.succeeded).toHaveLength(2);
    expect(
      res.resultsByPlugin.python.succeeded[0].scanResult.identity.targetFile,
    ).toEqual('dev.txt');
    expect(
      res.resultsByPlugin.python.succeeded[1].scanResult.identity.targetFile,
    ).toEqual('prod.txt');
  });
  it.only('Snyk fix returns results as expected when 1 fails to fix', async () => {
    jest.spyOn(fs, 'readFileSync').mockImplementation((fileName) => {
      if (fileName === 'dev.txt') {
        throw new Error('Invalid encoding');
      } else {
        return JSON.stringify({ django: { version: '^1.9.0', line: 3 } });
      }
    });
    const txtProdProjectTestResult = generateEntityToFix('pip', 'prod.txt');
    const txtDevProjectTestResult = generateEntityToFix('pip', 'dev.txt');
    const pipfileProjectTestResult = generateEntityToFix('pip', 'Pipfile');
    const res = await snykFix.fix([
      txtDevProjectTestResult,
      txtProdProjectTestResult,
      pipfileProjectTestResult,
    ]);
    expect(res.exceptionsByScanType).toEqual({});
    expect(Object.keys(res.resultsByPlugin)).toHaveLength(1);
    expect(Object.keys(res.resultsByPlugin)[0]).toEqual('python');
    // skipped unsupported
    expect(res.resultsByPlugin.python.skipped).toHaveLength(1);
    expect(res.resultsByPlugin.python.skipped[0]).toEqual(
      pipfileProjectTestResult,
    );

    // first *.txt throws because of the mock above
    expect(res.resultsByPlugin.python.failed).toHaveLength(1);
    expect(
      res.resultsByPlugin.python.failed[0].scanResult.identity.targetFile,
    ).toEqual('dev.txt');

    expect(res.resultsByPlugin.python.succeeded).toHaveLength(1);

    expect(
      res.resultsByPlugin.python.succeeded[0].scanResult.identity.targetFile,
    ).toEqual('prod.txt');
  });
});

describe('groupEntitiesPerScanType', () => {
  it('It correctly groups related entities per handler type (pip)', () => {
    const txtProdProjectTestResult = generateEntityToFix('pip', 'prod.txt');
    const txtDevProjectTestResult = generateEntityToFix('pip', 'dev.txt');
    const pipfileProjectTestResult = generateEntityToFix('pip', 'Pipfile');
    const res = snykFix.groupEntitiesPerScanType([
      txtProdProjectTestResult,
      txtDevProjectTestResult,
      pipfileProjectTestResult,
    ]);
    expect(Object.keys(res)[0]).toEqual('pip');
    expect(Object.keys(res)[0]).toHaveLength(3);
  });
  it('It correctly groups related entities per handler type (mixed)', () => {
    const txtProdProjectTestResult = generateEntityToFix('pip', 'prod.txt');
    const txtDevProjectTestResult = generateEntityToFix('pip', 'dev.txt');
    const npmProjectTestResult = generateEntityToFix('npm', 'package.json');
    const res = snykFix.groupEntitiesPerScanType([
      txtProdProjectTestResult,
      txtDevProjectTestResult,
      npmProjectTestResult,
    ]);
    expect(Object.keys(res).sort()).toEqual(['npm', 'pip']);
    expect(res.npm).toHaveLength(1);
    expect(res.pip).toHaveLength(2);
  });

  it('It correctly groups related entities per handler type with missing type', () => {
    const txtProdProjectTestResult = generateEntityToFix('pip', 'prod.txt');
    const txtDevProjectTestResult = generateEntityToFix('pip', 'dev.txt');
    const missingProjectTestResult = generateEntityToFix('npm', 'package.json');
    delete missingProjectTestResult.scanResult.identity.type;

    const res = snykFix.groupEntitiesPerScanType([
      txtProdProjectTestResult,
      txtDevProjectTestResult,
      missingProjectTestResult,
    ]);
    expect(Object.keys(res).sort()).toEqual(['missing-type', 'pip']);
    expect(res['missing-type']).toHaveLength(1);
    expect(res.pip).toHaveLength(2);
  });
});
