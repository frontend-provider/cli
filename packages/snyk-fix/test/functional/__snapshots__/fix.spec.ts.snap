// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snyk fix Snyk fix returns results for supported & unsupported type 1`] = `
Array [
  Object {
    "failed": Array [],
    "skipped": Array [
      Object {
        "scanResult": Object {
          "facts": Array [
            Object {
              "data": "not-implemented",
              "type": "not-implemented",
            },
          ],
          "identity": Object {
            "targetFile": "Pipfile",
            "type": "pip",
          },
        },
        "testResult": Object {
          "depGraphData": "",
          "issues": Array [
            Object {
              "fixInfo": Object {},
              "issueId": "VULN_ID_1",
              "pkgName": "package@version",
            },
          ],
          "issuesData": Object {
            "vuln-id": Object {
              "id": "VULN_ID_1",
              "severity": "high",
              "title": "Fake vuln",
            },
          },
        },
      },
    ],
    "succeeded": Array [
      Object {
        "scanResult": Object {
          "facts": Array [
            Object {
              "data": "not-implemented",
              "type": "not-implemented",
            },
          ],
          "identity": Object {
            "targetFile": "requirements.txt",
            "type": "pip",
          },
        },
        "testResult": Object {
          "depGraphData": "",
          "issues": Array [
            Object {
              "fixInfo": Object {},
              "issueId": "VULN_ID_1",
              "pkgName": "package@version",
            },
          ],
          "issuesData": Object {
            "vuln-id": Object {
              "id": "VULN_ID_1",
              "severity": "high",
              "title": "Fake vuln",
            },
          },
        },
      },
    ],
  },
]
`;

exports[`Snyk fix Snyk fix returns results for supported type 1`] = `
Object {
  "exceptionsByScanType": Object {},
  "resultsByPlugin": Object {
    "requirements.txt": Object {
      "failed": Array [],
      "skipped": Array [],
      "succeeded": Array [
        Object {
          "scanResult": Object {
            "facts": Array [
              Object {
                "data": "not-implemented",
                "type": "not-implemented",
              },
            ],
            "identity": Object {
              "targetFile": "requirements.txt",
              "type": "pip",
            },
          },
          "testResult": Object {
            "depGraphData": "",
            "issues": Array [
              Object {
                "fixInfo": Object {},
                "issueId": "VULN_ID_1",
                "pkgName": "package@version",
              },
            ],
            "issuesData": Object {
              "vuln-id": Object {
                "id": "VULN_ID_1",
                "severity": "high",
                "title": "Fake vuln",
              },
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Snyk fix Snyk fix throws error when called with unsupported type 1`] = `"Error G10 (UnsupportedTypeError) npm is not supported."`;
